### ISSUE TITLE
Question about validate function

### ISSUE THREAD

**Issue by michiel-j** [raymin0223/patch-mix_contrastive_learning#9]
Hi @raymin0223,

Thank you for making your research code open source.

I have a question about the `validate` function in `main.py` file. I think I do not fully understand the working of the `validate` function.

I do fully understand the first half of this function, but after the calculation of specificity (`sp`), sensitivity (`se`) and ICBHI score (`sc`), I might be missing something. I think these metrics are calculated for every batch of the validation set. But then the validation metrics of the final batch are printed to the console. I would expect to see metrics printed in the console that are averaged out across all validation batches? Would this reasoning be correct?

The `validate` function currently looks like this:
```python
def validate(val_loader, model, classifier, criterion, args, best_acc, best_model=None):
    save_bool = False
    model.eval()
    classifier.eval()

    batch_time = AverageMeter()
    losses = AverageMeter()
    top1 = AverageMeter()
    hits, counts = [0.0] * args.n_cls, [0.0] * args.n_cls

    with torch.no_grad():
        end = time.time()
        for idx, (images, labels, metadata) in enumerate(val_loader):
            images = images.cuda(non_blocking=True)
            labels = labels.cuda(non_blocking=True)
            bsz = labels.shape[0]

            with torch.cuda.amp.autocast():
                features = model(images)
                output = classifier(features)
                loss = criterion[0](output, labels)

            losses.update(loss.item(), bsz)
            [acc1], _ = accuracy(output, labels, topk=(1,))
            top1.update(acc1[0], bsz)

            _, preds = torch.max(output, 1)
            for idx in range(preds.shape[0]):
                counts[labels[idx].item()] += 1.0
                if not args.two_cls_eval:
                    if preds[idx].item() == labels[idx].item():
                        hits[labels[idx].item()] += 1.0
                else:  # only when args.n_cls == 4
                    if labels[idx].item() == 0 and preds[idx].item() == labels[idx].item():
                        hits[labels[idx].item()] += 1.0
                    elif labels[idx].item() != 0 and preds[idx].item() > 0:  # abnormal
                        hits[labels[idx].item()] += 1.0

            sp, se, sc = get_score(hits, counts)

            batch_time.update(time.time() - end)
            end = time.time()

            if (idx + 1) % args.print_freq == 0:
                print('Test: [{0}/{1}]\t'
                      'Time {batch_time.val:.3f} ({batch_time.avg:.3f})\t'
                      'Loss {loss.val:.4f} ({loss.avg:.4f})\t'
                      'Acc@1 {top1.val:.3f} ({top1.avg:.3f})'.format(
                       idx + 1, len(val_loader), batch_time=batch_time,
                       loss=losses, top1=top1))
    
    if sc > best_acc[-1] and se > 5:
        save_bool = True
        best_acc = [sp, se, sc]
        best_model = [deepcopy(model.state_dict()), deepcopy(classifier.state_dict())]

    print(' * S_p: {:.2f}, S_e: {:.2f}, Score: {:.2f} (Best S_p: {:.2f}, S_e: {:.2f}, Score: {:.2f})'.format(sp, se, sc, best_acc[0], best_acc[1], best_acc[-1]))
    print(' * Acc@1 {top1.avg:.2f}'.format(top1=top1))

    return best_acc, best_model, save_bool
```

But in my reasoning, it would have to look more or less like this:
```python
def validate(val_loader, model, classifier, criterion, args, best_acc, best_model=None):
    save_bool = False
    model.eval()
    classifier.eval()

    batch_time = AverageMeter()
    losses = AverageMeter()
    top1 = AverageMeter()
    hits, counts = [0.0] * args.n_cls, [0.0] * args.n_cls

    # Create AverageMeter for all metrics, such that these can be averaged out across all validation batches
    specificities = AverageMeter()
    sensitivities = AverageMeter()
    scores = AverageMeter()

    with torch.no_grad():
        end = time.time()
        for idx, (images, labels, metadata) in enumerate(val_loader):
            images = images.cuda(non_blocking=True)
            labels = labels.cuda(non_blocking=True)
            bsz = labels.shape[0]

            with torch.cuda.amp.autocast():
                features = model(images)
                output = classifier(features)
                loss = criterion[0](output, labels)

            losses.update(loss.item(), bsz)
            [acc1], _ = accuracy(output, labels, topk=(1,))
            top1.update(acc1[0], bsz)

            _, preds = torch.max(output, 1)
            for idx in range(preds.shape[0]):
                counts[labels[idx].item()] += 1.0
                if not args.two_cls_eval:
                    if preds[idx].item() == labels[idx].item():
                        hits[labels[idx].item()] += 1.0
                else:  # only when args.n_cls == 4
                    if labels[idx].item() == 0 and preds[idx].item() == labels[idx].item():
                        hits[labels[idx].item()] += 1.0
                    elif labels[idx].item() != 0 and preds[idx].item() > 0:  # abnormal
                        hits[labels[idx].item()] += 1.0

            sp, se, sc = get_score(hits, counts)

            # Add metrics of this validation batch to the AverageMeter of each metric
            specificities.update(sp)
            sensitivities.update(se)
            scores.update(sc)

            batch_time.update(time.time() - end)
            end = time.time()

            if (idx + 1) % args.print_freq == 0:
                print('Test: [{0}/{1}]\t'
                      'Time {batch_time.val:.3f} ({batch_time.avg:.3f})\t'
                      'Loss {loss.val:.4f} ({loss.avg:.4f})\t'
                      'Acc@1 {top1.val:.3f} ({top1.avg:.3f})'.format(
                       idx + 1, len(val_loader), batch_time=batch_time,
                       loss=losses, top1=top1))
    
    # Save model if the average ICBHI score is better than the current best one
    if scores.avg > best_acc[-1] and sensitivities.avg > 5:
        save_bool = True
        best_acc = [specificities.avg, sensitivities.avg, scores.avg]
        best_model = [deepcopy(model.state_dict()), deepcopy(classifier.state_dict())]

    print(' * S_p: {:.2f}, S_e: {:.2f}, Score: {:.2f} (Best S_p: {:.2f}, S_e: {:.2f}, Score: {:.2f})'.format(specificities.avg, sensitivities.avg, scores.avg, best_acc[0], best_acc[1], best_acc[-1]))
    print(' * Acc@1 {top1.avg:.2f}'.format(top1=top1))

    return best_acc, best_model, save_bool
```

Could you please help me figure out where my reasoning might be wrong?

Thank you very much.

**Comment by raymin0223** [comment]
Hi @michiel-j,

So sorry for the late reply -- idk why I missed this issues and notification emails..

Thanks for reporting some suspicious codes. And to be honest, I should have implemented much clear haha..
Your intuition is perfect, and I'll explain why I implemented it in that way, and what it could be wrong.

Evaluation of all samples in validation datasets will be stored `hits` and `counts` at the end of the for loop for val_loader.
Thus, at the end of the loop, the last `sp, se, sc` will be the final metrics for validation datasets.
I'm assuming it is correct implementation to get the final evaluation metrics, intermediate metrics for three metrics (sp, se, sc) are kinda confusing though.

And about your suggested codes, not sure, but it could induce different metrics of sp, se, and sc. Since AverageMeter will continuously average the values after the updates at each step.

**Comment by michiel-j** [comment]
Hi @raymin0223 

Thank you for your answer.
No need to apologise for the late reply.

I understand the function working now. Your `validate` function behaves correctly indeed.

I will close this issue now.
